envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -68,12 +68,12 @@ class SnapshotUpdater(
                         result.groups
                     }
 
-                    if (result.adsSnapshot != null && result.xdsSnapshot != null) {
+                    if (result.adsSnapshot != null || result.xdsSnapshot != null) {
                         versions.retainGroups(cache.groups())
                         groups.forEach { group ->
-                            if (group.ads) {
+                            if (result.adsSnapshot != null && properties.configurationMode.ads && group.ads) {
                                 updateSnapshotForGroup(group, result.adsSnapshot)
-                            } else {
+                            } else if (result.xdsSnapshot != null && properties.configurationMode.xds && !group.ads) {
                                 updateSnapshotForGroup(group, result.xdsSnapshot)
                             }
```

Comment:
Maybe add: `else { log.error("Requested snapshot for ${group.ads ? "ads" : "xds"} mode, but it is not here. This should never happen")  }`
It should never happen, so we have to know if it happens despite that :)